pageserver/src/http/routes.rs

```diff
@@ -0,0 +1,164 @@
+            &request_data.name,
+            &request_data.start_point,
+            &request_data.tenant_id,
+        )
+    })
+    .await
+    .map_err(ApiError::from_err)??;
+    Ok(json_response(StatusCode::CREATED, response_data)?)
+}
+
+async fn branch_list_handler(request: Request<Body>) -> Result<Response<Body>, ApiError> {
+    let tenantid: ZTenantId = match request.param("tenant_id") {
+        Some(arg) => arg
+            .parse()
+            .map_err(|_| ApiError::BadRequest("failed to parse tenant id".to_string()))?,
+        None => {
+            return Err(ApiError::BadRequest(
+                "no tenant id specified in path param".to_string(),
+            ))
+        }
+    };
+
+    check_permission(&request, Some(tenantid))?;
+
+    let state = request.data::<Arc<State>>().unwrap().clone();
+    let response_data =
+        tokio::task::spawn_blocking(move || crate::branches::get_branches(state.conf, &tenantid))
+            .await
+            .map_err(ApiError::from_err)??;
+    Ok(json_response(StatusCode::OK, response_data)?)
+}
+
+async fn tenant_list_handler(request: Request<Body>) -> Result<Response<Body>, ApiError> {
+    // check for management permission
+    check_permission(&request, None)?;
+
+    let state = request.data::<Arc<State>>().unwrap().clone();
+    let response_data =
+        tokio::task::spawn_blocking(move || crate::branches::get_tenants(state.conf))
+            .await
+            .map_err(ApiError::from_err)??;
+    Ok(json_response(StatusCode::OK, response_data)?)
+}
+
+async fn tenant_create_handler(mut request: Request<Body>) -> Result<Response<Body>, ApiError> {
+    // check for management permission
+    check_permission(&request, None)?;
+
+    let state = request.data::<Arc<State>>().unwrap().clone();
+    let request_data: TenantCreateRequest = json_request(&mut request).await?;
+
+    let response_data = tokio::task::spawn_blocking(move || {
+        page_cache::create_repository_for_tenant(state.conf, request_data.tenant_id)
+    })
+    .await
+    .map_err(ApiError::from_err)??;
+    Ok(json_response(StatusCode::CREATED, response_data)?)
+}
+
+async fn handler_404(_: Request<Body>) -> Result<Response<Body>, ApiError> {
+    json_response(
+        StatusCode::NOT_FOUND,
+        HttpErrorBody::from_msg("page not found".to_owned()),
+    )
+}
+
+pub fn get_router(
+    conf: &'static PageServerConf,
+    auth: Arc<Option<JwtAuth>>,
+) -> RouterBuilder<hyper::Body, ApiError> {
+    let spec = include_bytes!("openapi_spec.yml");
+    let mut router = attach_openapi_ui(endpoint::get_router(), spec, "/swagger.yml", "/v1/doc");
+    if let Some(_) = &auth.as_ref() {
+        // note that State is used as a type parameteer without an Arc
+        // this is a simple solution because it is not possible to implement
+        // AuthProvider for Arc<State> so middleware assumes that state is wrapped in Arc
+        router = router.middleware(Middleware::pre(auth_middleware::<State>))
```

Comment:
The thing that bothers me most is `auth_middleware::<State>`. It feels fragile. I'll post an update after I've come up with an alternative solution.